napari-workshops/SciPy-0722/make_a_simple_plugin.md

```diff
@@ -47,95 +49,141 @@ You will be asked for some information to customize the setup of your plugin. Ea
+
+    ```python
+    @magic_factory
+    def example_magic_widget(img_layer: "napari.layers.Image"):
+        print(f"you have selected {img_layer}")
+    ```
+
+	- Finally, you see the what looks like just a plain function. We will modify this to incorporate our `detect_spots` function.
 
 	```python
-	# This is the actual plugin function, where we export our function
-	# (The functions themselves are defined below)
-	@napari_hook_implementation
-	def napari_experimental_provide_function():
-	    # we can return a single function
-	    # or a tuple of (function, magicgui_options)
-	    # or a list of multiple functions with or without options, as shown here:
-	    return [threshold, image_arithmetic]
+    # Uses the `autogenerate: true` flag in the plugin manifest
+    # to indicate it should be wrapped as a magicgui to autogenerate
+    # a widget.
+    def example_function_widget(img_layer: "napari.layers.Image"):
+        print(f"you have selected {img_layer}")
 	```
+
+    - Find the `Command` ID in `napari.yaml` that points to `example_function_widget`, and then find that `Command` ID in the
+`Widgets` contribution section. Note that unlike the other `widget` contributions, this one includes `autogenerate: true`.
+
+    ```yaml
+    - command: napari-spot-detector.make_func_widget
+      autogenerate: true
+      display_name: Example Function Widget
+    ```
+
+    - This means our function doesn't need to know anything about magicgui. If we provide type annotations to the parameters,
+the GUI widgets will be generated for us without even a decorator!
 	
-	- Finally, you see two example functions. We will replace these with our `detect_spots()` function.
-6. Delete the example functions. You can delete everything below the comment `# 1.  First example,...`.
-7. Copy the `gaussian_high_pass()` and `detect_spots()` functions from your notebook from the first part of the tutorial and paste it where the example functions were (the ones you deleted in the previous step).
-8. Next, we need to modify `detect_spots()` to return the necessary layer data so that napari can create a new Points layer with our detected spots. If `detect_spots()` returns a `LayerDataTuple`, napari will add a new layer to the viewer using the data in the `LayerDataTuple`. For more information on the `LayerDataTuple` type, please see the lecture slides or [the docs](https://napari.org/guides/stable/magicgui.html#returning-napari-types-layerdatatuple).
-	- The layer data tuple should be: `(layer_data, layer_options, layer_type)`
+	- Let's edit `example_function_widget` to do our spot detection.
```

Comment:
I think it would be good to pull this out to a "numbered heading" level so that it's more obvious. I think it's easy to miss that we are using this approach and it helps to know to provide context for the following instructions.

It might be nice to add a bit of rationale as well. Something like "we are using the function widget approach so that we do not need to program the GUI ourselves".